pkg/filetree/filetree.go

```diff
@@ -0,0 +1,754 @@
+package filetree
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/anchore/stereoscope/internal"
+	"github.com/anchore/stereoscope/pkg/file"
+	"github.com/anchore/stereoscope/pkg/filetree/filenode"
+	"github.com/anchore/stereoscope/pkg/tree"
+	"github.com/anchore/stereoscope/pkg/tree/node"
+	"github.com/bmatcuk/doublestar/v2"
+)
+
+var ErrRemovingRoot = errors.New("cannot remove the root path (`/`) from the FileTree")
+var ErrLinkCycleDetected = errors.New("cycle during symlink resolution")
+
+// FileTree represents a file/directory Tree
+type FileTree struct {
+	tree *tree.Tree
+}
+
+// NewFileTree creates a new FileTree instance.
+func NewFileTree() *FileTree {
+	t := tree.NewTree()
+
+	// Initialize FileTree with a root "/" Node
+	_ = t.AddRoot(filenode.NewDir("/", nil))
+
+	return &FileTree{
+		tree: t,
+	}
+}
+
+// Copy returns a Copy of the current FileTree.
+func (t *FileTree) Copy() (*FileTree, error) {
+	ct := NewFileTree()
+	ct.tree = t.tree.Copy()
+	return ct, nil
+}
+
+// AllFiles returns all files and directories within the FileTree.
+func (t *FileTree) AllFiles() []file.Reference {
+	var files []file.Reference
+	for _, n := range t.tree.Nodes() {
+		f := n.(*filenode.FileNode)
+		if f.FileType == file.TypeReg && f.Reference != nil {
+			files = append(files, *f.Reference)
+		}
+	}
+	return files
+}
+
+func (t *FileTree) AllRealPaths() []file.Path {
+	var files []file.Path
+	for _, n := range t.tree.Nodes() {
+		f := n.(*filenode.FileNode)
+		if f != nil {
+			files = append(files, f.RealPath)
+		}
+	}
+	return files
+}
+
+func (t *FileTree) ListPaths(dir file.Path) ([]file.Path, error) {
+	n, err := t.node(dir, linkResolutionStrategy{
+		FollowAncestorLinks: true,
+		FollowBasenameLinks: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if n == nil {
+		return nil, nil
+	}
+
+	if n.FileType != file.TypeDir {
+		return nil, nil
+	}
```

Comment:
wouldn't it be easier for the consumer of this method to get back an empty slice of Path (`[].file.Path`) instead of a `nil`? The method definition here is about listing paths, and from what I can see how this is consumed, the consumers aren't checking if they get a `nil` back

Reply:
this is one of the nice features of go, where a `nil` slice is the same as an empty slice (`[]file.Path{}` or `make([]file.Path, 0)`)